File: Project2/lib/deck.rb
# Class for Deck Object
class Deck
  attr_accessor :deck_of_cards, :current_deck, :cards_selected, :displayed_sprites, :index_array, :selected_index
  attr_reader :numbers

  @@numbers = %w[one two three]
  @@shape = %w[diamond squiggle oval]
  @@texture = %w[solid striped open]
  @@color = %w[red green purple]

  def initialize
    # array of undealt cards
    @deck_of_cards = []
    # array of dealt cards
    @current_deck = []
    # array of cards selected by player
    @cards_selected = []
    # array of sprites displayed to player
    @displayed_sprites = []
    # array containing the index of the card objects selected
    @index_array = []
    @selected_index = -1
  end

  # Creates a deck of 81 cards. Returns a deck of 81 cards containing one of each card in the game
  def create_card_deck
    (0...@@numbers.length).each do |i|
      (0...@@shape.length).each do |j|
        (0...@@texture.length).each do |k|
          (0...@@color.length).each do |l|
            # location of the card sprite in file
            image_path = "images/#{@@color[l]}_#{@@numbers[i]}_#{@@shape[j]}_#{@@texture[k]}.png"
            # creates a card with each of the 4 properties
            @deck_of_cards.push(Cards::SetCard.new(@@color[l], @@numbers[i], @@shape[j], @@texture[k], image_path))
          end
        end
      end
    end
  end

  # clears the current selection of cards
  def clear_selected
    @index_array.clear
    @cards_selected.clear
  end

  # Checks if the array of 3 cards is a set.
  # Parameter cards: list of 3 cards to determine if the 3 are a set
  # Return true/false depending on it cards form a valid set
  def is_set?(cards)
    color = false
    number = false
    shape = false
    texture = false

    # check if all colors are unique OR all colors are the same
    if (cards[0].color == cards[1].color && cards[1].color ==
        cards[2].color && cards[0].color == cards[2].color) ||
       (cards[0].color != cards[1].color && cards[1].color !=
       cards[2].color && cards[0].color != cards[2].color)
      color = true
    end

    # check if all numbers are unique OR all numbers are the same
    if (cards[0].number == cards[1].number && cards[1].number ==
        cards[2].number && cards[0].number == cards[2].number) ||
       (cards[0].number != cards[1].number && cards[1].number !=
       cards[2].number && cards[0].number != cards[2].number)
      number = true
    end

    # check if all shape are unique OR all shaoe are the same
    if (cards[0].shape == cards[1].shape && cards[1].shape ==
        cards[2].shape && cards[0].shape == cards[2].shape) ||
       (cards[0].shape != cards[1].shape && cards[1].shape !=
       cards[2].shape && cards[0].shape != cards[2].shape)
      shape = true
    end

    # check if all texture are unique OR all texture are the same
    if (cards[0].texture == cards[1].texture && cards[1].texture ==
        cards[2].texture && cards[0].texture == cards[2].texture) ||
       (cards[0].texture != cards[1].texture && cards[1].texture !=
       cards[2].texture && cards[0].texture != cards[2].texture)
      texture = true
    end

    # return true if each attribute is unique amonst cards OR the same amongst cards within that category
    color && number && shape && texture
  end

  # Function to check if deck of cards is empty (game ends)
  # Parameter deck: the current deck of cards for which to check the length
  # Return true/false if deck is empty/not
  def deck_empty?(deck)
    deck.length.zero?
  end

  # Function to find if there are any sets remaining in the deck
  # Parameter deck: card deck to determine if set is remaining in deck
  # Returns true/false if there is set remaining/not
  def set_remaining?(deck)
    found_set = false
    i = 0
    while i < deck.length && !found_set
      card1 = deck[i]
      j = i + 1
      while j < deck.length && !found_set
        card2 = deck[j]
        k = j + 1
        while k < deck.length && !found_set
          card3 = deck[k]
          found_set = is_set?([card1, card2, card3])
          k += 1
        end
        j += 1
      end
      i += 1
    end
    found_set
  end

  # keeps track of what cards are selected
  def add_to_selected
    card = @current_deck[@selected_index]
    puts "Card #{@selected_index + 1} has already been selected" if @cards_selected.include? card
    # if card is not selected by player previously, add card
    unless @cards_selected.include? card
      puts "Card #{@selected_index + 1} selected"
      @index_array << selected_index
      @cards_selected << card
    end
  end

  # Determines which card is clicked by the player and perform the click behavior
  # Parameter card_deck: current deck of cards from which to select from
  # Parameter event: GUI event to determine if a particular card is clicked
  # Parameter player: the current player who is selecting
  def select_cards(card_deck, event, player)
    # checks which card is selected by player
    @displayed_sprites.each_index do |i|
      # define dimensions of the card
      card = @current_deck[i]
      card.width = card.x + 100
      card.height = card.y + 56
      @selected_index = i
      # actions when the card sprite is clicked
      card.clicked_behavior(card_deck, player) if card.clicked? event
    end
  end

  # Function to replace with new cards in current deck
  def replace_card
    @index_array.each_index do |i|
      index = @index_array[i].to_i
      # get a new card
      if !deck_empty?(@deck_of_cards) && (@current_deck.length <= 12)
        card = @deck_of_cards.pop
        # set x & y of new card with the x & y of old card
        card.x = @current_deck[index].x
        card.y = @current_deck[index].y
        # replace previous card
        @current_deck[index] = card
        GUI.replace_sprite(card, self, index)
      else
        # If that there are more than 12 cards or deck of cards is empty just delete cards
        # Set the card at the index to nil
        @current_deck[index] = nil
      end
    end
    # Delete all nil objects (the cards to be deleted) from current deck
    @current_deck.compact!
    # Display the updated deck to GUI
    Cards.print_cards(self)
  end
end

File: lib/javascript/deck.js
// Class for Deck Object
class Deck {
  constructor() {
    this.numbers = ["one", "two", "three"];
    this.shape = ["diamond", "squiggle", "oval"];
    this.texture = ["solid", "striped", "open"];
    this.color = ["red", "green", "purple"];
    this.deck_of_cards = [];
    this.current_deck = [];
    this.cards_selected = [];
    this.displayed_cards = [];
    this.array_index = [];
    this.selected_index = -1;
  }

  /* clears the current selection of cards */
  create_card_deck() {
    for (let i = 0; i < this.numbers.length; i++) {
      for (let j = 0; j < this.shape.length; j++) {
        for (let k = 0; k < this.texture.length; k++) {
          for (let l = 0; l < this.color.length; l++) {
            let image_path = `lib/images/${this.color[l]}_${this.numbers[i]}_${this.shape[j]}_${this.texture[k]}.png`;
            this.deck_of_cards.push(
              new Cards(
                this.color[l],
                this.numbers[i],
                this.shape[j],
                this.texture[k],
                image_path
              )
            );
          }
        }
      }
    }
  }

  // Checks if the array of 3 cards is a set.
  is_set(cards) {
    let color = false;
    let number = false;
    let shape = false;
    let texture = false;

    /* check if all colors are unique OR all colors are the same */
    if (
      (cards[0].color == cards[1].color &&
        cards[1].color == cards[2].color &&
        cards[0].color == cards[2].color) ||
      (cards[0].color != cards[1].color &&
        cards[1].color != cards[2].color &&
        cards[0].color != cards[2].color)
    ) {
      color = true;
    }

    /* check if all numbers are unique OR all numbers are the same */
    if (
      (cards[0].number == cards[1].number &&
        cards[1].number == cards[2].number &&
        cards[0].number == cards[2].number) ||
      (cards[0].number != cards[1].number &&
        cards[1].number != cards[2].number &&
        cards[0].number != cards[2].number)
    ) {
      number = true;
    }

    /* check if all shape are unique OR all shaoe are the same */
    if (
      (cards[0].shape == cards[1].shape &&
        cards[1].shape == cards[2].shape &&
        cards[0].shape == cards[2].shape) ||
      (cards[0].shape != cards[1].shape &&
        cards[1].shape != cards[2].shape &&
        cards[0].shape != cards[2].shape)
    ) {
      shape = true;
    }

    /* check if all texture are unique OR all texture are the same */
    if (
      (cards[0].texture == cards[1].texture &&
        cards[1].texture == cards[2].texture &&
        cards[0].texture == cards[2].texture) ||
      (cards[0].texture != cards[1].texture &&
        cards[1].texture != cards[2].texture &&
        cards[0].texture != cards[2].texture)
    ) {
      texture = true;
    }

    /* return true if each attribute is unique amonst cards OR the same amongst cards within that category */
    return color && number && shape && texture;
  }

  /* Function to check if deck of cards is empty (game ends)
    Parameter deck: the current deck of cards for which to check the length
    Return true/false if deck is empty/not */
  deck_empty(deck) {
    return deck.length == 0;
  }

  /* Function to find if there are any sets remaining in the deck
    Parameter deck: card deck to determine if set is remaining in deck
    Returns true/false if there is set remaining/not */
  set_remaining(deck) {
    let found_set = false;
    let i = 0;
    while (i < deck.length && !found_set) {
      let card1 = deck[i];
      let j = i + 1;
      while (j < deck.length && !found_set) {
        let card2 = deck[j];
        let k = j + 1;
        while (k < deck.length && !found_set) {
          let card3 = deck[k];
          found_set = this.is_set([card1, card2, card3]);
          k += 1;
        }
        j += 1;
      }
      i += 1;
    }
    return found_set;
  }

  //  Determines which card is clicked by the player and perform the click behavior
  select_cards(elem) {
    /* checks which card is selected by player */
    let index = parseInt(elem.id) - 1;
    this.displayed_cards.forEach((i) => {
      let card = this.current_deck[index];
      this.selected_index = index;
      //Check if card has been clicked by checking border property
      let elem = document.getElementById((index + 1).toString());
      if (elem != null && elem.classList.contains("card-selected")) {
        card.clicked_behavior();
      }
      index++;
    });
  }

  /* Function to replace with new cards in current deck */
  replace_card() {
    let card = new Cards();
    this.array_index.forEach((i) => {
      let index = parseInt(i);
      /* get a new card */
      if (
        !this.deck_empty(this.deck_of_cards) &&
        this.current_deck.length <= 12
      ) {
        let card = this.deck_of_cards.pop();
        /* replace previous card */
        this.current_deck[index] = card;
        card.create_card_sprite(this, index);
      } else {
        /* If that there are more than 12 cards or deck of cards is empty just delete cards
          Set the card at the index to nil */
        this.current_deck[index] = null;
        this.displayed_cards[index] = null;
      }
    });
    /* Delete all nil objects (the cards to be deleted) from current deck */
    this.current_deck = this.current_deck.filter((n) => n);
    this.displayed_cards = this.displayed_cards.filter((n) => n);
    /* Display the updated deck to GUI */
    card.print_cards(this);
  }

  //add card to selected deck of cards if selected
  add_to_selected() {
    let c = card_deck.current_deck[card_deck.selected_index];
    if (card_deck.cards_selected.indexOf(c) == -1) {
      //Push the card index into index array
      card_deck.array_index.push(card_deck.selected_index);
      //Push card into cards selected arrayW
      card_deck.cards_selected.push(c);
      c.selected = true;
    }
  }
}
